File: luna-studio/node-editor/js/websocket.js
"use strict";

var mockSocket = function () {
  return {
    addEventListener: function () {},
    removeEventListener: function () {},
    send: function () {}
  };
};

module.exports = function () {
  var listenConnection = mockSocket();
  var sendConnection   = mockSocket();
  var isConnOpen = false;
  var listeners = {
    onOpen: [],
    onMessage: [],
    onClose: [],
    onError: []
  };

  var attachListeners = function () {
    listeners.onOpen.forEach(function (listener) {
      listenConnection.addEventListener("open", listener);
    });
    listeners.onMessage.forEach(function (listener) {
      listenConnection.addEventListener("message", listener);
    });
    listeners.onClose.forEach(function (listener) {
      listenConnection.addEventListener("close", listener);
    });
    listeners.onError.forEach(function (listener) {
      listenConnection.addEventListener("error", listener);
    });
  };

  return {
    isOpen: function () {
      return isConnOpen;
    },
    onOpen: function (listener) {
      isConnOpen = true;
      listeners.onOpen.push(listener);
      listenConnection.addEventListener("open", listener);
    },
    onMessage: function (listener) {
      listeners.onMessage.push(listener);
      listenConnection.addEventListener("message", listener);
    },
    onClose: function (listener) {
      listeners.onClose.push(listener);
      listenConnection.addEventListener("close", listener);
    },
    onError: function (listener) {
      listeners.onError.push(listener);
      listenConnection.addEventListener("error", listener);
    },
    connect: function (listenAddr, sendAddr) {
      listenConnection = new WebSocket(listenAddr);
      listenConnection.binaryType = "arraybuffer";
      sendConnection = new WebSocket(sendAddr);
      sendConnection.binaryType = "arraybuffer"
      attachListeners();
    },
    send: function (data) {
      sendConnection.send(data);
    }
  };
};

File: luna-studio/lib/src/WebSocket.hs
{-# LANGUAGE JavaScriptFFI #-}

module WebSocket (WebSocket
                 , connect
                 , getCode
                 , getData
                 , getWebSocket
                 , isOpen
                 , onClose
                 , onError
                 , onMessage
                 , onOpen
                 , send
                 ) where

import           Data.ByteString                   (ByteString)
import           Data.JSString
import           GHCJS.Foreign.Callback
import           GHCJS.Marshal.Pure                ( PFromJSVal (..)
                                                   , PToJSVal (..))
import           GHCJS.Types                       (IsJSVal)
import           GHCJS.Buffer                      ( toByteString
                                                   , createFromArrayBuffer
                                                   , fromByteString
                                                   , getArrayBuffer)
import           JavaScript.TypedArray.ArrayBuffer (ArrayBuffer)
import           Common.Prelude

newtype WebSocket      = WebSocket      JSVal deriving (PFromJSVal, PToJSVal)
newtype WSMessageEvent = WSMessageEvent JSVal deriving (PFromJSVal, PToJSVal)
newtype WSClosedEvent  = WSClosedEvent  JSVal deriving (PFromJSVal, PToJSVal)

instance IsJSVal WebSocket
instance IsJSVal WSMessageEvent

foreign import javascript safe "$1.data"
    js_getData :: WSMessageEvent -> IO ArrayBuffer

foreign import javascript safe "$1.code"
    getCode :: WSClosedEvent -> IO Int

foreign import javascript safe "init.websocket"
    getWebSocket :: IO WebSocket


foreign import javascript safe "$1.isOpen()"
    isOpen :: WebSocket -> IO Bool

foreign import javascript safe "$1.onOpen($2)"
    onOpen' :: WebSocket -> Callback (IO ()) -> IO ()

foreign import javascript safe "$1.unOnOpen($2)"
    unOnOpen' :: WebSocket -> Callback (IO ()) -> IO ()

foreign import javascript safe "$1.onMessage($2)"
    onMessage' :: WebSocket -> Callback (JSVal -> IO ()) -> IO ()

foreign import javascript safe "$1.unOnMessage()"
    unOnMessage' :: WebSocket -> Callback (JSVal -> IO ()) -> IO ()

foreign import javascript safe "$1.onClose($2)"
    onClose' :: WebSocket -> Callback (JSVal -> IO ()) -> IO ()

foreign import javascript safe "$1.unOnClose()"
    unOnClose' :: WebSocket -> Callback (JSVal -> IO ()) -> IO ()

foreign import javascript safe "$1.onError($2)"
    onError' :: WebSocket -> Callback (IO ()) -> IO ()

foreign import javascript safe "$1.unOnError()"
    unOnError' :: WebSocket -> Callback (IO ()) -> IO ()

foreign import javascript safe "$1.send($2)"
    js_send :: WebSocket -> ArrayBuffer -> IO ()

foreign import javascript safe "$1.connect($2, $3)"
    connect' :: WebSocket -> JSString -> JSString -> IO ()

connect :: WebSocket -> String -> String -> IO ()
connect ws listenAddr sendAddr = connect' ws (convert listenAddr)
                                             (convert sendAddr)

getData :: WSMessageEvent -> IO ByteString
getData = fmap (toByteString 0 Nothing . createFromArrayBuffer) . js_getData

send :: WebSocket -> ByteString -> IO ()
send ws = js_send ws . getArrayBuffer . view _1 . fromByteString

onOpen :: WebSocket -> IO () -> IO (IO ())
onOpen ws callback = do
    wrappedCallback <- asyncCallback callback
    onOpen' ws wrappedCallback
    return $ unOnOpen' ws wrappedCallback >> releaseCallback wrappedCallback

onMessage :: WebSocket -> (WSMessageEvent -> IO ()) -> IO (IO ())
onMessage ws callback = do
    wrappedCallback <- asyncCallback1 (callback . pFromJSVal)
    onMessage' ws wrappedCallback
    return $ unOnMessage' ws wrappedCallback >> releaseCallback wrappedCallback

onClose :: WebSocket -> (WSClosedEvent -> IO ()) -> IO (IO ())
onClose ws callback = do
    wrappedCallback <- asyncCallback1 (callback . pFromJSVal)
    onClose' ws wrappedCallback
    return $ unOnClose' ws wrappedCallback >> releaseCallback wrappedCallback

onError :: WebSocket -> IO () -> IO (IO ())
onError ws callback = do
    wrappedCallback <- asyncCallback callback
    onError' ws wrappedCallback
    return $ unOnError' ws wrappedCallback >> releaseCallback wrappedCallback
